### ISSUE TITLE
Need howto on Kopia portable and updated docs

### ISSUE THREAD

**Issue by githubbapoopa** [kopia/kopia#2282]
My goal  is to have a portable version of kopia (and rclone) on Win10.

So I started by downloading and unpacking https://github.com/kopia/kopia/releases/download/v0.11.3/KopiaUI-0.11.3-win.zip
Also I downloaded and unpacked https://github.com/rclone/rclone/releases/download/v1.59.0/rclone-v1.59.0-windows-amd64.zip in the same directory.

Was able to setup kopia and rclone to use a Dropbox repository sucessfully. Snapshots also work. So far so good.

But KopiaUI-0.11.3-win despite several issues here:

https://github.com/kopia/kopia/issues/368
https://github.com/kopia/kopia/pull/398
https://github.com/kopia/kopia/issues/428
https://github.com/kopia/kopia/pull/431
https://github.com/kopia/kopia/issues/367

still seems to defaut to --config-file=C:\Users\user\AppData\Roaming\kopia\repository.config which I don*t want for my portable solution.

Any used config file should ideally not be outside the portable folder. What do I have to do to change that?

Noob question: How to pass the command line switch --user-data-dir='full\path' to KopiaUI? For example let's say I want to use the folder d:\kopia portabe\user data

Is it also possible to use a relative path for the config file (making kopia truly portable while using it on a stick etc.)?

Next step would the be to get rclone also portable which raises the same questions (pass command line for paths / relative paths possible?)

Any help much apprectiated. I would also encourage you to maybe write a quick tutorial within your docs.

Apologies if this described in detail anywhere else yet. Couldn't find it...

At first I couldn't believe the speed of the snapshot and thought of an error. But no kopia is *that* fast! Wow.

Thx a lot

**Comment by basldfalksjdf** [comment]
Reading [#3](https://github.com/kopia/kopia/pull/398), it says `Portable mode is enabled by creating 'repositories' subdirectory before launching the app.` Did you create the folder? If that does not work, I am not sure if KopiaUI currently supports custom location of the config file. If you read https://github.com/kopia/kopia/pull/398, you will see that the `repositories` folder mentioned in that PR only stores cache and logs, not the config file. 

If you look at the screenshots at https://github.com/kopia/kopia/issues/428, it actually looks like custom location of config file was a feature before but has since been removed. If you also read https://github.com/kopia/kopia/pull/1374, looks like there was some plan to have a feature that converts KopiaUI to portable mode but the feature has not yet been implemented.

Long winded way to saying... I don't think KopiaUI currently supports this. But I do agree that this would be a great feature to add. 

If you want portability, you can achieve it using Kopia CLI. Create a repo, copy the config file to your portable folder, and connect to the repo using `kopia repository connect from-config` command: https://kopia.io/docs/reference/command-line/common/repository-connect-from-config/

**Comment by githubbapoopa** [comment]
Thanks for looking deeply into that. I did create the "repositories" folder inside my portable folder. When I then configure the repository (rclone in my case, see later) KopiaUI then stores the repository.config inside that "repositories" folder. So far so good. 

This repository.config file sets the cache to a relative path to that "repositories" folder (in my case KopiaUI portable\"repositories"\cache\repository\)

So if I undertsnad it right, this way I coul set a custom cache and logs dir via the repository.config file (either absolute or relative path)?


This is my repository.config

{
  "storage": {
    "type": "rclone",
    "config": {
      "remotePath": "dropbox:kopia",
      "rcloneExe": "d:\\_____Trash\\KopiaUI portable\\rclone x64 portable\\rclone.exe",
      "atomicWrites": false,
      "dirShards": null
    }
  },
  "caching": {
    "cacheDirectory": "cache\\repository",
    "maxCacheSize": 5242880000,
    "maxMetadataCacheSize": 5242880000,
    "maxListCacheDuration": 30
  },
  "hostname": "myhostname",
  "username": "myusername",
  "description": "My Repository",
  "enableActions": false,
  "formatBlobCacheDuration": 900000000000


  As you can see above things are slightly more complicated as I also use rclone for accessing Dropbox.
  This breaks portability a bit in my case as rclone itself again uses \AppData\Local\Temp\ in Win.

  So the very dump question is: How to use these "flags" for kopia or rclone? Is is a simple commandline? Please give me a quick headsup here. I know just dumb...

  Rclone doc states you can use commandline arguments or environment variables

  https://rclone.org/install/ > Portable installation (command-line arguments)
  https://rclone.org/docs/#environment-variables

   Do I get it right that I could modify the repository.config to use rclone *with* it's "flags" to set rclone temp dir and rclone cache dir? Is it as simple as:

     "rcloneExe": "d:\\_____Trash\\KopiaUI portable\\rclone x64 portable\\rclone.exe --cache-dir=x:\cache --temp-dir=x:\temp"
?

  This way kopia would use its own cache and temp dir via its repository.conf file and within that file I could also set rclone to use its own cache and temp dir. This would be ideal for portability.

  thx

**Comment by basldfalksjdf** [comment]
`"rcloneExe": "d:\\_____Trash\\KopiaUI portable\\rclone x64 portable\\rclone.exe --cache-dir=x:\cache --temp-dir=x:\temp"` is not the official way to pass arguments to rclone. It may or may not work.

The official way you can pass arguments to rclone (such as `--cache-dir=1 and 1--temp-dir=`) is by using the `--rclone-args` option with `kopia repository create rclone` command or `kopia repository connect rclone` command. See https://kopia.io/docs/reference/command-line/common/repository-create-rclone/ or https://kopia.io/docs/reference/command-line/common/repository-connect-rclone/.

However, this is available only via Kopia CLI. KopiaUI does not yet (officially) support the ability to pass arguments to rclone. There may potential be an unofficial way for you to pass arguments to rclone when using KopiaUI. I have never tried this myself and it isnt officially supported, so it may not work, but you may want to give it a try. 

You see, KopiaUI is really only a wrapper for Kopia CLI. So, in theory, if you can make KopiaUI's repository.config mimic the repository.config created by Kopia CLI, you may be able to achieve what you want. This may or may not work:

(1) Using Kopia CLI to create an rclone repository using the `kopia repository create rclone --rclone-exe= --remote-path= --rclone-args="--cache-dir= and --temp-dir="` command

(2) Open the repository.config created by Kopia CLI and copy the line associated with the `--rclone-args` option.

(3) Paste that line in the repository.confg for KopiaUI.

(4) Restart KopiaUI

(5) Maybe profit?

**Comment by githubbapoopa** [comment]
Thanks for pointing me in the right direction here. Via your explanations I was able to identify the parameters for rclone.

This is my folder structure (for reference). I tried to make every folder seperate for understanding the structure.

d:\PortableApps\KopiaUI portable\
d:\PortableApps\KopiaUI portable\rclone x64 portable\
d:\Temp\kopia
d:\Temp\kopia\cache
d:\Temp\kopia\temp
d:\Temp\rclone
d:\Temp\rclone\cache
d:\Temp\rclone\temp


For getting the right/different config lines you indeed have to first set up the repository vie kopia.exe

Replace/insert these lines in d:\PortableApps\KopiaUI portable\repositories\repository.config

    "config": {
      "remotePath": "dropbox:kopia",
      "rcloneExe": "d:\\PortableApps\\KopiaUI portable\\rclone x64 portable\\rclone.exe",
      "rcloneArgs": [
        "--cache-dir=d:\\Temp\\rclone\\cache and --temp-dir=\\Temp\\rclone\\temp"
      ],


This config seems to work for rclone.

However I now got the problem upside down as I am not able to set the kopia chache dir as d:\Temp\kopia\cache

I tried the same steps via kopia.exe to fimd the right place in the repository.config file

I replaced

  },
  "caching": {
    "cacheDirectory": "cache\\repository",
    "maxCacheSize": 5242880000,
    "maxMetadataCacheSize": 5242880000,
    "maxListCacheDuration": 30
  },
  
  with

  },
  "caching": {
    "cacheDirectory": "d:\\Temp\\kopia\\cache",
    "maxCacheSize": 5242880000,
    "maxMetadataCacheSize": 5242880000,
    "maxListCacheDuration": 30
  },

  But no matter what, KopiaUI always uses d:\PortableApps\KopiaUI portable\repositories\cache\repository

  Is this a limitation of KopiaUI?

  Halfway there...
